File: iOS-HR/HRManager/Source/MVC/Model/Event.swift
//
//  Event.swift
//  HRManager
//
//  Created by Tinh Vu on 1/13/19.
//  Copyright © 2019 Tinh Vu. All rights reserved.
//

import Foundation
import ObjectMapper

class Event : ServiceModel {
    var leaveGroup : Int = 0
    var leaveTypeId : Int = 0
    var leaveTypeName : String?
    var leaveOfAbsenceType : String?
    var des : String?
    var startDate : String?
    var endDate : String?
    var countHour : Int?
    var countDay : Float?
    var registerUserId : String?
    var approveUserId : String?
    var status : String?
    var reseanRefusal : String?
    var comment : String?
    var cCEmail : String?
    var leaveRemaining : Int?
    var thoiGianNghi : Int = 0
    var listWorkflow : String?
    var registerId : String?
    var createdOnDate : String?
    var lastModifiedOnDate : String?
    var publishOnDate : String?
    var transitionTime : String?
    var createdByUser : User?
    var lastModifiedByUser : User?
    var registerUser : String?
    var approveUser : String?
    var publishByUser : String?
    var transitionUser : User?
    var workflowCode : String?
    var workflowState : String?
    var allowedToEmployeeNames : String?
    var workflowPreviousState : String?
    var registerCategoryCode: String?

    var state: WorkFlow {
        guard let stateString = workflowState else {
            return .Draft
        }
        return WorkFlow(rawValue: stateString) ?? .Draft
    }

    var createdBy: String {
        return "\(createdByUser?.fullName ?? "") - \(registerCategoryCode ?? "")"
    }

    var createdDate: Date? {
        guard let dateString = startDate else {
            return nil
        }
        return dateString.toDate()
    }

    var endedDate: Date? {
        guard let endString = endDate else {
            return nil
        }
        return endString.toDate()
    }

    required init?(map: Map) {
        super.init(map: map)
    }

    override func mapping(map: Map) {

        leaveGroup <- map["LeaveGroup"]
        leaveTypeId <- map["LeaveTypeId"]
        leaveTypeName <- map["LeaveTypeName"]
        leaveOfAbsenceType <- map["LeaveOfAbsenceType"]
        des <- map["Description"]
        startDate <- map["StartDate"]
        endDate <- map["EndDate"]
        countHour <- map["CountHour"]
        countDay <- map["CountDay"]
        registerUserId <- map["RegisterUserId"]
        approveUserId <- map["ApproveUserId"]
        status <- map["Status"]
        reseanRefusal <- map["ReseanRefusal"]
        comment <- map["Comment"]
        cCEmail <- map["CCEmail"]
        leaveRemaining <- map["LeaveRemaining"]
        thoiGianNghi <- map["ThoiGianNghi"]
        listWorkflow <- map["ListWorkflow"]
        registerId <- map["RegisterId"]
        createdOnDate <- map["CreatedOnDate"]
        lastModifiedOnDate <- map["LastModifiedOnDate"]
        publishOnDate <- map["PublishOnDate"]
        transitionTime <- map["TransitionTime"]
        createdByUser <- map["CreatedByUser"]
        lastModifiedByUser <- map["LastModifiedByUser"]
        registerUser <- map["RegisterUser"]
        approveUser <- map["ApproveUser"]
        publishByUser <- map["PublishByUser"]
        transitionUser <- map["TransitionUser"]
        workflowCode <- map["WorkflowCode"]
        workflowState <- map["WorkflowState"]
        allowedToEmployeeNames <- map["AllowedToEmployeeNames"]
        workflowPreviousState <- map["WorkflowPreviousState"]
        registerCategoryCode <- map["RegisterCategoryCode"]
    }

}

class User: ServiceModel {
    var applicationId : String?
    var pwd : String?
    var active : Bool?
    var roles : String?
    var rights : String?
    var newPwd : String?
    var isExpire : Bool?
    var userId : String?
    var userName : String?
    var fullName : String?
    var position : String?
    var mobile : String?
    var staffCode : String?
    var staffImg: String?

    required init?(map: Map) {
        super.init(map: map)
    }

    override func mapping(map: Map) {

        applicationId <- map["ApplicationId"]
        pwd <- map["Pwd"]
        active <- map["Active"]
        roles <- map["Roles"]
        rights <- map["Rights"]
        newPwd <- map["newPwd"]
        isExpire <- map["isExpire"]
        userId <- map["UserId"]
        userName <- map["UserName"]
        fullName <- map["FullName"]
        position <- map["Position"]
        mobile <- map["Mobile"]
        staffCode <- map["StaffCode"]
        staffImg <- map["StaffImg"]
    }
}

File: tinhvv-android-hr-3ed1b6f37f8e/app/src/main/java/com/athsoftware/hrm/model/Event.kt
package com.athsoftware.hrm.model

import com.athsoftware.hrm.App
import com.athsoftware.hrm.helper.extensions.applyOn
import com.athsoftware.hrm.network.APIClient
import com.google.gson.annotations.SerializedName
import io.reactivex.disposables.Disposable
import java.io.Serializable
import java.lang.ref.WeakReference

data class Event(
        @SerializedName("LeaveGroup")
        var leaveGroup: Int? = null,
        @SerializedName("LeaveTypeId")
        var leaveTypeId: Int? = null,
        @SerializedName("LeaveTypeName")
        var leaveTypeName: String? = null,
        @SerializedName("LeaveOfAbsenceType")
        var leaveOfAbsenceType: String? = null,
        @SerializedName("Description")
        var des: String? = null,
        @SerializedName("StartDate")
        var startDate: String? = null,
        @SerializedName("EndDate")
        var endDate: String? = null,
        @SerializedName("CountHour")
        var countHour: Float? = null,
        @SerializedName("CountDay")
        var countDay: Float? = null,
        @SerializedName("RegisterUserId")
        var registerUserId: String? = null,
        @SerializedName("ApproveUserId")
        var approveUserId: String? = null,
        @SerializedName("Status")
        var status: String? = null,
        @SerializedName("ReseanRefusal")
        var reseanRefusal: String? = null,
        @SerializedName("Comment")
        var comment: String? = null,
        @SerializedName("CCEmail")
        var cCEmail: String? = null,
        @SerializedName("LeaveRemaining")
        var leaveRemaining: Float? = null,
        @SerializedName("ThoiGianNghi")
        var thoiGianNghi: Float? = null,
        @SerializedName("ListWorkflow")
        var listWorkflow: String? = null,
        @SerializedName("RegisterId")
        var registerId: String? = null,
        @SerializedName("CreatedOnDate")
        var createdOnDate: String? = null,
        @SerializedName("LastModifiedOnDate")
        var lastModifiedOnDate: String? = null,
        @SerializedName("PublishOnDate")
        var publishOnDate: String? = null,
        @SerializedName("TransitionTime")
        var transitionTime: String? = null,
        @SerializedName("CreatedByUser")
        var createdByUser: User? = null,
        @SerializedName("LastModifiedByUser")
        var lastModifiedByUser: User? = null,
        @SerializedName("RegisterUser")
        var registerUser: User? = null,
        @SerializedName("ApproveUser")
        var approveUser: User? = null,
        @SerializedName("TransitionUser")
        var transitionUser: User? = null,
        @SerializedName("WorkflowCode")
        var workflowCode: String? = null,
        @SerializedName("WorkflowState")
        var workflowState: String? = null,
        @SerializedName("AllowedToEmployeeNames")
        var allowedToEmployeeNames: String? = null,
        @SerializedName("WorkflowPreviousState")
        var workflowPreviousState: String? = null,
        @SerializedName("RegisterCategoryCode")
        var registerCategoryCode: String? = null
) : Serializable {
    var disposable: Disposable? = null
    fun loadUserImage(position: Int, listener: (String, Int) -> Unit) {
        val weakListener = WeakReference(listener)
        val userId = createdByUser?.userId ?: kotlin.run {
            weakListener.get()?.invoke("", position)
            return
        }
        val image = App.shared().imageCache[userId]
        if (image != null) {
            weakListener.get()?.invoke(image, position)
            return
        }
        if (App.shared().imageCacheListener[userId] == null) {
            App.shared().imageCacheListener[userId] = mutableListOf(weakListener)
            disposable = APIClient.shared.api.getProfile(userId)
                    .applyOn()
                    .subscribe({
                        val userImage = it.data?.staffImg ?: ""
                        App.shared().imageCache[userId] = userImage
                        App.shared().imageCacheListener[userId]?.forEach {
                            it.get()?.invoke(userImage, position)
                        }
                    }, {
                        App.shared().imageCacheListener[userId]?.forEach {
                            it.get()?.invoke("", position)
                        }
                    })
        } else {
            App.shared().imageCacheListener[userId]!!.add(weakListener)
        }
    }

    var createName: String = ""
        get() = String.format("%s - %s", createdByUser?.fullName ?: "", registerCategoryCode ?: "")
}